x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriodTests.java

```diff
@@ -0,0 +1,92 @@
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.AbstractFunctionTestCase;
+import org.elasticsearch.xpack.esql.expression.function.FunctionName;
+import org.elasticsearch.xpack.esql.expression.function.TestCaseSupplier;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+@FunctionName("to_dateperiod")
+public class ToDatePeriodTests extends AbstractFunctionTestCase {
+    public ToDatePeriodTests(@Name("TestCase") Supplier<TestCaseSupplier.TestCase> testCaseSupplier) {
+        this.testCase = testCaseSupplier.get();
+    }
+
+    @ParametersFactory
+    public static Iterable<Object[]> parameters() {
+        // TODO multivalue fields
+        String read = "Attribute[channel=0]";
+        List<TestCaseSupplier> suppliers = new ArrayList<>();
+
+        TestCaseSupplier.unary(
+            suppliers,
+            read,
+            validDatePeriod(DataType.DATE_PERIOD),
+            DataType.DATE_PERIOD,
+            bytesRef -> EsqlDataTypeConverter.parseTemporalAmount(((BytesRef) bytesRef).utf8ToString(), DataType.DATE_PERIOD),
+            List.of()
+        );
+
+        for (DataType inputType : AbstractConvertFunction.STRING_TYPES) {
+            TestCaseSupplier.unary(
+                suppliers,
+                read,
+                validDatePeriod(inputType),
+                DataType.DATE_PERIOD,
+                bytesRef -> EsqlDataTypeConverter.parseTemporalAmount(((BytesRef) bytesRef).utf8ToString(), DataType.DATE_PERIOD),
+                List.of()
+            );
+        }
+
+        return parameterSuppliersFromTypedData(
+            errorsForCasesWithoutExamples(anyNullIsNull(true, suppliers), (v, p) -> "date_period or string")
+        );
+    }
+
+    @Override
+    protected Expression build(Source source, List<Expression> args) {
+        return new ToDatePeriod(source, args.get(0));
+    }
+
+    private static List<TestCaseSupplier.TypedDataSupplier> validDatePeriod(DataType dataType) {
+        return List.of(
+            new TestCaseSupplier.TypedDataSupplier("1 day", () -> new BytesRef("1 day"), dataType),
+            new TestCaseSupplier.TypedDataSupplier("1 d", () -> new BytesRef("1 d"), dataType),
+            new TestCaseSupplier.TypedDataSupplier("10 days", () -> new BytesRef("10 days"), dataType),
+
+            new TestCaseSupplier.TypedDataSupplier("1 week", () -> new BytesRef("1 week"), dataType),
+            new TestCaseSupplier.TypedDataSupplier("1 w", () -> new BytesRef("1 w"), dataType),
+            new TestCaseSupplier.TypedDataSupplier("5 weeks", () -> new BytesRef("5 weeks"), dataType),
```

Comment:
Here, I think we should also not duplicate all the suffixes for periods, because this can go out of sync easily.

I think we should re-use the same list of "week" suffixes that I suggested we use in `EsqlDataTypeConverter#parseTemporalAmount`. We should also probably assert that the total number of different suffixes we test here is the same as the total number of period suffixes that are used in `parseTemporalAmount`.

Similarly for durations.